### PR TITLE
Added interruptable reentrant locks instead of synchronized block for batching of messages. Suppressed stack track for scale-down calls

### PR DESCRIPTION
**Pull request by vdantu**
# Description of changes:
1. Added interruptable locks instead of synchronized. This helps in retrying of killed backend workers.
2. Suppressed exceptions dumped when scaling down.
3. Renamed the state to SCALED_DOWN so that its clear in the logs.
4. Removed unnecessary states.

# Testing

## With reentrant locks (Changes in this PR)
| Concurrency | TPS | P50 | P90 | P99 | P100 |
| ---------        | ---  | ---   | ---   | ---   | ---- |
| 5 | 801.64 | 6 | 6 | 9 | 18 |
| 8  | 843.44 | 11 | 12 | 14 | 41 |
| 10 | 1672.37 | 6 | 6 | 8 | 27 |
| 25 | 4206.87 | 6 | 6 | 9 | 29 |
| 50 | 6370.77 | 7 | 9 | 55 | 63 |

## With sychronized block (Changes already in  master)
| Concurrency | TPS | P50 | P90 | P99 | P100 |
| ---------        | ---  | ---   | ---   | ---   | ---- |
| 5 | 810.55 | 6 | 6 | 8 | 37 |
| 8  | 840.93 | 11 | 12 | 15 | 26 |
| 10 | 1649.30 | 6 | 6 | 9 | 12 |
| 25 | 4225.86 | 6 | 6 | 9 | 16 |
| 50 | 6163.47 | 8 | 10 | 31 | 39 |

**To run CI tests on your changes refer [README.md](https://github.com/awslabs/mxnet-model-server/blob/master/ci/README.md)**

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/Model.java [hunk 1]
```diff
@@ -34,6 +35,7 @@
     private int maxWorkers;
     private int batchSize;
     private int maxBatchDelay;
+    ReentrantLock lock;
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/Model.java]: Should this be private?

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 2]
```diff
@@ -304,22 +310,26 @@ public void shutdown() {
     }
 
     void setState(WorkerState newState) {
+        logger.debug("State {} -> {}", state, newState);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: This log is redundant, there is a log in listener.notifyChangeState() already

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 3]
```diff
@@ -304,22 +310,26 @@ public void shutdown() {
...
             this.state = newState;
         }
-
         if (state == WorkerState.WORKER_MODEL_LOADED) {
             workerLoadTime.setValue(String.valueOf(timeTaken));
             workerLoadTime.setTimestamp(
                     String.valueOf(TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis())));
             loggerMmsMetrics.info(workerLoadTime);
         }
+        logger.debug("State changed to {}", state);
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: This log is redundant:
1. you have a log at the beginning already
2. there is a log in listener.notifyChangeState() already

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java [hunk 4]
```diff
@@ -304,22 +310,26 @@ public void shutdown() {
     }
 
     void setState(WorkerState newState) {
+        logger.debug("State {} -> {}", state, newState);
         listener.notifyChangeState(model.getModelName(), newState);
         long timeTaken = System.currentTimeMillis() - startTime;
-        if (state != WorkerState.WORKER_TERMINATED) {
+        if ((state != WorkerState.WORKER_SCALED_DOWN)
+                || (newState == WorkerState.WORKER_SCALED_DOWN)) {
```
Comment (frankfliu) [inline frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: Not sure what you are doing here.
What you added case is state == WorkerState.WORKER_SCALED_DOWN and newState == WorkerState.WORKER_SCALED_DOWN
What's the point to assign the newState?

**Comment by frankfliu** [comment]
You commit message is too long.

A git commit message should follow:

First line of commit message should less than 80 chars.
Then empty line,
After empty line, you can add whatever message you want.

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java (hunk 4 above)
Comment (vdantu) [reply frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: You are correct. This is a redundant check.

File: frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java (hunk 3 above)
Comment (vdantu) [reply frontend/server/src/main/java/com/amazonaws/ml/mms/wlm/WorkerThread.java]: I don't think we capture state transition .. We only capture what it is set to, but its hard to track what it was before.

**Comment by vdantu** [comment]
@frankfliu , addressed your comments.

**Pull request merged**